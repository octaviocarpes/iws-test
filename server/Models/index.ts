export { Album } from './Album';
export { Band } from './Band';
