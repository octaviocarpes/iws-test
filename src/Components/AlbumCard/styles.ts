import { createUseStyles } from 'react-jss';
import { colors } from '../../styles/colors';

export const useStyles = createUseStyles({
  cardWrapper: {
    width: '50%',
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
  },

  albumImage: {
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
    paddingLeft: 10,
    paddingRight: 10,
  },

  name: {
    fontWeight: 'bold',
    color: colors.primaryGray,
    whiteSpace: 'nowrap',
  },

  albumInfo: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-around',
    alignItems: 'left',
  },
});
