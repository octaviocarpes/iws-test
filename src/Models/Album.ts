export interface Album {
  name: string;
  image: string;
  releasedDate: Date;
  band: string;
  tracks: [
    {
      name: string;
      duration: number;
      id: string;
    },
  ];
  id: string;
}
