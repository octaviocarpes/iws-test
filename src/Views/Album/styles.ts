import { createUseStyles } from 'react-jss';
import { colors } from '../../styles/colors';

export const useStyles = createUseStyles({
  albumView: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    width: '100%',
    height: '100%',
    fontFamily: 'Montserrat',
  },

  goBack: {
    position: 'fixed',
    top: 0,
    left: 0,
    width: 100,
    height: 40,
    marginTop: 15,
    marginLeft: 15,

    '& button': {
      width: '100%',
      height: '100%',
      background: 'white',
      fontSize: 16,
      fontFamily: 'Montserrat',
      fontWeight: 'bold',
      color: colors.primaryOrange,
      border: [`1px solid ${colors.primaryOrange}`],
      borderRadius: 20,

      '&:focus': {
        outline: 'none',
      },
    },
  },

  albumLogo: {
    border: ['1px solid black'],
    width: '100%',
    height: 400,

    '& img': {
      width: '100%',
      height: 400,
    },
  },

  albumHeader: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',

    '& h1': {
      color: colors.primaryGray,
    },

    '& h3': {
      color: colors.primaryOrange,
      textTransform: 'capitalize',
    },
  },

  biography: {
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 15,

    '& p': {
      color: colors.primaryGray,
      textAlign: 'justify',
      lineHeight: 1.8,
      fontSize: 15,
    },

    '& a': {
      color: colors.primaryOrange,
    },
  },

  tracks: {
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',

    '& h2': {
      color: colors.primaryOrange,
    },
  },

  tracksWrapper: {
    width: '90%',
    paddingLeft: 24,
    paddingRight: 24,
    paddingBottom: 10,
  },

  track: {
    marginBottom: 15,
  },
});
