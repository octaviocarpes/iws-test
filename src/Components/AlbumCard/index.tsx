import React, { memo, ReactElement } from 'react';
import { Album } from '../../Models/Album';
import { useStyles } from './styles';
import fallbackImage from '../../styles/img/album-icon.png';

import Image from '../Image';
import Card from '../Card';

interface Props {
  album: Album;
  onTouch?: () => void;
}

const AlbumCard = memo(
  (props: Props): ReactElement => {
    const { album, onTouch } = props;
    const styles = useStyles();

    const _handleTouchEnd = (): void => {
      if (onTouch) onTouch();
    };

    return (
      <>
        <Card onTouch={(): void => _handleTouchEnd()}>
          <div className={styles.cardWrapper}>
            <div className={styles.albumImage}>
              <Image alt={`${album.name}-image`} src={album.image} fallbackSrc={fallbackImage} />
            </div>
            <div className={styles.albumInfo}>
              <p className={styles.name}>{album.name}</p>
            </div>
          </div>
        </Card>
      </>
    );
  },
);

AlbumCard.displayName = 'AlbumCard';
export default AlbumCard;
