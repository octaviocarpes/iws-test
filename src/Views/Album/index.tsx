import React, { useState, useEffect, useContext } from 'react';
import { observer } from 'mobx-react';
import { AxiosResponse } from 'axios';

import { LoadingContext } from '../../Contexts/Loading.context';
import { getAlbumById } from '../../Services/AlbumsService';
import { Album } from '../../Models/Album';

import { useStyles } from './styles';
import fallbackImage from '../../styles/img/album-icon.png';

import Image from '../../Components/Image';

const AlbumView = observer(props => {
  const loadingStore = useContext(LoadingContext);
  const [album, setAlbum] = useState<Album>();
  const styles = useStyles();

  useEffect(() => {
    if (album === undefined) {
      loadingStore.setIsLoading(true);
      getAlbumById(props.id)
        .then((response: AxiosResponse) => {
          setAlbum(response.data);
          loadingStore.setIsLoading(false);
        })
        .catch(error => {
          // TODO: notify error,
          console.log(error);
          loadingStore.setIsLoading(false);
        });
    }
    // eslint-ignore-next-line
  }, [album, loadingStore, props]);

  const calculateDuration = (time: number): string => {
    const minutes = Math.floor(time / 60);
    const seconds = time - minutes * 60;
    return `${minutes}:${seconds}`;
  };

  return (
    <>
      {album !== undefined ? (
        <div className={styles.albumView}>
          <div className={styles.goBack}>
            <button onClick={(): void => window.history.back()}>Go Back</button>
          </div>
          <div className={styles.albumLogo}>
            <Image src={album.image} fallbackSrc={fallbackImage} alt="album Image" />
          </div>
          <div className={styles.albumHeader}>
            <h1>{album.name}</h1>
          </div>
          <div className={styles.biography}>
            <p>Release date: {new Date(album.releasedDate).toLocaleDateString()}</p>
          </div>
          <div className={styles.tracks}>
            <h2>Tracks</h2>
            <div className={styles.tracksWrapper}>
              {album.tracks.map(track => (
                <div className={styles.track} key={track.id}>
                  <p>
                    {track.name} - Duration: {calculateDuration(track.duration)}
                  </p>
                </div>
              ))}
            </div>
          </div>
        </div>
      ) : (
        <></>
      )}
    </>
  );
});

export default AlbumView;
