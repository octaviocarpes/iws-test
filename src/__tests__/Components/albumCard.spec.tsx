import React from 'react';
import { render, cleanup } from '@testing-library/react';
import AlbumCard from '../../Components/AlbumCard';
import { Album } from '../../Models/Album';

const album: Album = {
  name: 'Album',
  band: 'Band',
  id: '12345',
  image: 'image',
  releasedDate: new Date(),
  tracks: [
    {
      id: '',
      duration: 180,
      name: 'Tack',
    },
  ],
};

afterEach(cleanup);

describe('<AlbumCard />', () => {
  it('should contains the albums name', () => {
    const { getByText } = render(<AlbumCard album={album} />);
    getByText('Album');
  });

  it('should render the component correctly', () => {
    const tree = render(<AlbumCard album={album} />);
    expect(tree).toMatchSnapshot();
  });
});
