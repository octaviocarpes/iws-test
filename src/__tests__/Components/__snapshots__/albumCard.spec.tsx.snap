// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlbumCard /> should render the component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card-0-2-5"
      >
        <div
          class="cardWrapper-0-2-1"
        >
          <div
            class="albumImage-0-2-2"
          >
            <img
              alt="Album-image"
              src="image"
            />
          </div>
          <div
            class="albumInfo-0-2-4"
          >
            <p
              class="name-0-2-3"
            >
              Album
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card-0-2-5"
    >
      <div
        class="cardWrapper-0-2-1"
      >
        <div
          class="albumImage-0-2-2"
        >
          <img
            alt="Album-image"
            src="image"
          />
        </div>
        <div
          class="albumInfo-0-2-4"
        >
          <p
            class="name-0-2-3"
          >
            Album
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
